### PR TITLE
Enable the use of pre-trained models

### PR DESCRIPTION
**Pull request by Dexterp37**
Initial commit, not ready for review and not working.

### CODE REVIEW

File: profondo/cnn.py [hunk 1]
```diff
@@ -82,6 +84,103 @@ def build_model(x_train, y_train):
...
+
+    if not exists(TOP_MODEL_WEIGHTS_FILE):
+        # Build the "bottleneck-features" to feed into the top layers, to init
+        # the output layers with reasonable features: we basically record the
+        # output of the last convolutional block with our data and feed that
+        # into the new top block.
+        # Note: the "rescale" part is required by the pre-trained networks. We
+        # don't shuffle the samples to be able to pick their labels later on.
+        init_batch = 100
+        train_datagen = ImageDataGenerator(rescale=1. / 255)
```
Comment (vitillo) [inline profondo/cnn.py]: Our images are already in the [0, 1] range so they don't need rescaling.

**Review by vitillo — CHANGES_REQUESTED** [review]

File: profondo/cnn.py [hunk 2]
```diff
@@ -82,6 +84,102 @@ def build_model(x_train, y_train):
...
+    if not exists(TOP_MODEL_WEIGHTS_FILE):
+        # Build the "bottleneck-features" to feed into the top layers, to init
+        # the output layers with reasonable features: we basically record the
+        # output of the last convolutional block with our data and feed that
+        # into the new top block.
+        # Note: our images are already scaled in the [0, 1] range, we don't need
+        # the "rescale" part as required by the pre-trained networks. We
+        # don't shuffle the samples to be able to pick their labels later on.
+        init_batch = 100
+        train_datagen = ImageDataGenerator()
```
Comment (vitillo) [inline profondo/cnn.py]: Might be worthwhile removing the generator since we don't use it anymore.

**Review by vitillo — CHANGES_REQUESTED** [review]

File: profondo/cnn.py [hunk 3]
```diff
@@ -86,6 +94,109 @@ def build_model(x_train, y_train):
...
+                          optimizer=keras.optimizers.Adam())
+        logging.info(top_model.summary())
+        reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.2, patience=5,
+                                      min_lr=10e-6)
+        top_model.fit(bottleneck_train, bottleneck_train_labels,
+                      epochs=50,
+                      batch_size=init_batch,
+                      validation_data=(bottleneck_val, bottleneck_val_labels),
+                      class_weight=genre_weights,
+                      callbacks=[reduce_lr],
```
Comment (vitillo) [inline profondo/cnn.py]: Why did you remove the other callbacks?

**Review by vitillo — CHANGES_REQUESTED** [review]

File: profondo/cnn.py [hunk 4]
```diff
@@ -151,5 +251,6 @@ def report_callback(epoch, model, x_val, y_val):
         callbacks=callbacks)
 
     # Load best model and evaluate it on the test data
-    model = load_model(x_test, y_test)
-    report_callback(0, model, x_test, y_test)
+    if exists(MODEL_FILENAME):
```
Comment (vitillo) [inline profondo/cnn.py]: The model has to exist, if not it's fine to raise an exception.

File: profondo/cnn.py [hunk 5]
```diff
@@ -103,7 +202,7 @@ def load_model(x_train, y_train):
 
 def report_callback(epoch, model, x_val, y_val):
     if epoch % 5 == 0:
-        prediction_proba = model.predict_proba(x_val)
+        prediction_proba = model.predict(x_val)
```
Comment (vitillo) [inline profondo/cnn.py:200]: Why has this call been changed?

File: profondo/cnn.py [hunk 6]
```diff
@@ -126,21 +226,21 @@ def report_callback(epoch, model, x_val, y_val):
...
         EarlyStopping(monitor='val_loss', min_delta=0, patience=10, verbose=1, mode='auto'),
+        TensorBoard(log_dir=join(TB_DIR, 'full'), histogram_freq=1, write_graph=True, write_images=False),
         LambdaCallback(on_epoch_end=lambda e, l: report_callback(e, model, x_val, y_val))
     ]
 
     train_datagen = ImageDataGenerator()
     train_generator = train_datagen.flow(x_train, y_train, batch_size=args.batch_size)
 
-    model = load_model(x_train, y_train)
+    model = build_pre_trained(args.pre_trained, x_train, y_train, x_val, y_val)\
```
Comment (vitillo) [inline profondo/cnn.py]: We should use the non-pretrained model if no pre-trained network name is specified.

File: profondo/cnn.py [hunk 7]
```diff
@@ -86,6 +95,96 @@ def build_model(x_train, y_train):
...
+        # Build the "bottleneck-features" to feed into the top layers, to init
+        # the output layers with reasonable features: we basically record the
+        # output of the last convolutional block with our data and feed that
+        # into the new top block.
+        # Note: our images are already scaled in the [0, 1] range, we don't need
+        # the "rescale" part as required by the pre-trained networks. We
+        # don't shuffle the samples to be able to pick their labels later on.
+        init_batch = 100
+        num_train_batches = len(x_train) // init_batch
+        bottleneck_train = model.predict(x_train, num_train_batches, verbose=True)
```
Comment (vitillo) [inline profondo/cnn.py]: I don't think `num_train_batches` makes sense how it's defined and used when passed to the `batch_size` argument of `model.predict()`. Can we just pass `args.batch_size`?

File: profondo/cnn.py [hunk 8]
```diff
@@ -86,6 +95,96 @@ def build_model(x_train, y_train):
...
+        top_model.compile(loss=keras.losses.binary_crossentropy,
+                          optimizer=keras.optimizers.SGD(lr=0.01, momentum=0.9))
+        logging.info(top_model.summary())
+        callbacks = [
+            ReduceLROnPlateau(monitor='val_loss', factor=0.2, patience=5, min_lr=10e-6),
+            TensorBoard(log_dir=join(TB_DIR, 'top'), histogram_freq=1, write_graph=True, write_images=False)
+        ]
+        top_model.fit(bottleneck_train, y_train,
+                      epochs=50,
+                      batch_size=init_batch,
```
Comment (vitillo) [inline profondo/cnn.py]: Please use `args.batch_size`.

File: profondo/cnn.py [hunk 9]
```diff
@@ -86,6 +95,96 @@ def build_model(x_train, y_train):
...
+
+        genre_weights = compute_genre_weights(y_train)
+
+        # Compile and train the top_model with the "bottleneck-features".
+        top_model = generate_top_model(bottleneck_train.shape[1:], y_train.shape[1])
+        top_model.compile(loss=keras.losses.binary_crossentropy,
+                          optimizer=keras.optimizers.SGD(lr=0.01, momentum=0.9))
+        logging.info(top_model.summary())
+        callbacks = [
+            ReduceLROnPlateau(monitor='val_loss', factor=0.2, patience=5, min_lr=10e-6),
```
Comment (vitillo) [inline profondo/cnn.py:147]: Don't we want to save only the best model in case this ends up overfitting?

File: profondo/cnn.py [hunk 10]
```diff
@@ -86,6 +95,96 @@ def build_model(x_train, y_train):
...
+            TensorBoard(log_dir=join(TB_DIR, 'top'), histogram_freq=1, write_graph=True, write_images=False)
+        ]
+        top_model.fit(bottleneck_train, y_train,
+                      epochs=50,
```
Comment (vitillo) [inline profondo/cnn.py]: Please use `args.epochs`.

File: profondo/cnn.py [hunk 11]
```diff
@@ -86,6 +95,96 @@ def build_model(x_train, y_train):
...
+        layer.trainable = False
+
+    # Join the convolutional part and the top layers. See
+    # https://github.com/fchollet/keras/issues/4040#issuecomment-253309615
+    joint_model = Model(input=model.input, output=trained_top(model.output))
+
+    # compile the model with a SGD/momentum optimizer
+    # and a very slow learning rate.
+    joint_model.compile(loss=keras.losses.binary_crossentropy,
+                        optimizer=keras.optimizers.SGD(lr=1e-4, momentum=0.9))
```
Comment (vitillo) [inline profondo/cnn.py:178]: Any particular reason we are using SGD instead of Adam?

**Review by vitillo — CHANGES_REQUESTED** [review]

File: profondo/cnn.py (hunk 9 above)
Comment (Dexterp37) [reply profondo/cnn.py:147]: Good point, fixed.

File: profondo/cnn.py (hunk 5 above)
Comment (Dexterp37) [reply profondo/cnn.py:200]: `predict_proba` isn't defined for `Model` (see [here](https://keras.io/models/model/)) and it's just a legacy wrapper for `predict` in `Sequential` ([implementation](https://github.com/fchollet/keras/blob/master/keras/models.py#L982)).

File: profondo/cnn.py (hunk 10 above)
Comment (Dexterp37) [reply profondo/cnn.py]: Fixed!

File: profondo/cnn.py (hunk 8 above)
Comment (Dexterp37) [reply profondo/cnn.py]: Fixed!

File: profondo/cnn.py (hunk 11 above)
Comment (Dexterp37) [reply profondo/cnn.py:178]: It seems to work better than Adam when using a pre-trained network on my system (Adam diverges, while SGD seem to converge to a nicer optimum)

File: profondo/cnn.py (hunk 4 above)
Comment (Dexterp37) [reply profondo/cnn.py]: If the model doesn't exist, `load_model` builds a new one from scratch. I'm happy to leave this as it is anyway, as long as we're aware of that :)

Comment (vitillo) [reply profondo/cnn.py]: Right, but we really want to evaluate the best model on the test data there so I think it makes sense to assert on the existence of the model.

**Review by vitillo — APPROVED** [review]
LGTM

**Pull request merged**